pkg/utils/etcdutil/health_checker.go

```diff
@@ -137,46 +160,178 @@ func resetClientEndpoints(client *clientv3.Client, endpoints ...string) {
+		count   = checker.clientCount()
+		probeCh = make(chan healthProbe, count)
+		wg      sync.WaitGroup
 	)
 	checker.healthyClients.Range(func(key, value interface{}) bool {
 		wg.Add(1)
 		go func(key, value interface{}) {
 			defer wg.Done()
 			defer logutil.LogPanic()
 			var (
-				ep     = key.(string)
-				client = value.(*healthyClient)
+				ep          = key.(string)
+				healthyCli  = value.(*healthyClient)
+				client      = healthyCli.Client
+				healthState = healthyCli.healthState
+				latency     = healthyCli.latency
+				start       = time.Now()
 			)
-			if IsHealthy(ctx, client.Client) {
-				hch <- ep
-				checker.storeClient(ep, &healthyClient{
-					Client:     client.Client,
-					lastHealth: time.Now(),
-				})
+			// Check the health of the endpoint.
+			healthy := IsHealthy(ctx, client)
+			took := time.Since(start)
+			latency.Observe(took.Seconds())
+			if !healthy {
+				healthState.Set(0)
+				log.Warn("etcd endpoint is unhealthy",
+					zap.String("endpoint", ep),
+					zap.Duration("took", took),
+					zap.String("source", checker.source))
 				return
 			}
+			healthState.Set(1)
+			// If the endpoint is healthy, update its last health time.
+			checker.storeClient(ep, client, start)
+			// Send the healthy probe result to the channel.
+			probeCh <- healthProbe{ep, took}
 		}(key, value)
 		return true
 	})
 	wg.Wait()
-	close(hch)
-	for h := range hch {
-		healthyList = append(healthyList, h)
+	close(probeCh)
+	var (
+		lastEps   = checker.client.Endpoints()
+		pickedEps = checker.pickEps(probeCh)
+	)
+	if len(pickedEps) > 0 {
+		checker.updateEvictedEps(lastEps, pickedEps)
+		pickedEps = checker.filterEps(pickedEps)
+	}
+	return lastEps, pickedEps, !typeutil.AreStringSlicesEquivalent(lastEps, pickedEps)
+}
+
+// Divide the acceptable latency range into several parts, and pick the endpoints which
+// are in the first acceptable latency range. Currently, we only take the latency of the
+// last health check into consideration, and maybe in the future we could introduce more
+// factors to help improving the selection strategy.
+func (checker *healthChecker) pickEps(probeCh <-chan healthProbe) []string {
+	var (
+		count     = len(probeCh)
+		pickedEps = make([]string, 0, count)
+	)
+	if count == 0 {
+		return pickedEps
+	}
+	// Take the default value as an example, if we have 3 endpoints with latency like:
+	//   - A: 175ms
+	//   - B: 50ms
+	//   - C: 2.5s
+	// the distribution will be like:
+	//   - [0, 1s) -> {A, B}
+	//   - [1s, 2s)
+	//   - [2s, 3s) -> {C}
+	//   - ...
+	//  - [9s, 10s)
+	// Then the picked endpoints will be {A, B} and if C is in the last used endpoints, it will be evicted later.
+	factor := int(DefaultRequestTimeout / DefaultSlowRequestTime)
+	for i := 0; i < factor; i++ {
+		minLatency, maxLatency := DefaultSlowRequestTime*time.Duration(i), DefaultSlowRequestTime*time.Duration(i+1)
+		for probe := range probeCh {
+			if minLatency <= probe.took && probe.took < maxLatency {
```

Comment:
What will happen if probe.took less than the minLatency?

Reply:
`minLatency` starts from zero. So this would not happen.